Add JAVA template, flag to CODEMON

alphaX86 opened the pull request:
### Description:
* This PR resolves #14 except listening feature. 
* Added JAVA templates 
* To avoid confusion, flag arguments are added (example: -cpp, -java)
* Help is refined totally with added args, warnings and color coding for description
* Works perfectly and this is made as v0.0.1

**Solves Issue**: #14

alphaX86 committed:
Add JAVA template except listen and refactor CLI tool

alphaX86 commented on setup.py:
```diff
@@ -2,7 +2,7 @@
 setup(
   name='codemon',
-  version='0.0.1',
+  version='0.0.2',
```
Version is changed, to avoid confusion

ankushbhardwxj replied:
You can keep version changed locally for your personal use, please don't include it in the PR, since I don't want to update version of the package.

alphaX86 replied:
Ok sure. 1 minute...

ankushbhardwxj commented on codemon/CodemonHelp.py:
```diff
@@ -4,8 +4,9 @@ def showHelp():
-  print("codemon init -n <file>  - - - - - - - creates file with given name")
-  print("codemon listen  - - - - - - - - - - - compiles and gives output")
-  print("codemon practice <dirName>  - - - - - starts a practice session")
+  print(colored.red("NOTE: Use the options -cpp and -java correctly in place of <args> to work \n"))
```
Could you get rid of this line here ?

alphaX86 replied:
Done

ankushbhardwxj commented on codemon/CodemonHelp.py:
```diff
@@ -4,8 +4,9 @@ def showHelp():
-  print("codemon practice <dirName>  - - - - - starts a practice session")
+  print(colored.red("NOTE: Use the options -cpp and -java correctly in place of <args> to work \n"))
+  print(colored.yellow("codemon - - - - - - - - - - - - - - - shows help"))
+  print(colored.yellow("codemon init <args> <contestName>  - - - - - initialises a contest in CPP/Java format"))
```
rewrite it as : "initialises a contest directory using format given in options"

ankushbhardwxj replied:
Also, rewrite `<args>` as `<options>` and add it after `<contestName>`

ankushbhardwxj commented on codemon/CodemonHelp.py:
```diff
@@ -4,8 +4,9 @@ def showHelp():
+  print(colored.red("NOTE: Use the options -cpp and -java correctly in place of <args> to work \n"))
+  print(colored.yellow("codemon - - - - - - - - - - - - - - - shows help"))
+  print(colored.yellow("codemon init <args> <contestName>  - - - - - initialises a contest in CPP/Java format"))
+  print(colored.yellow("codemon init <args> -n <file>  - - - - - - - creates file with given name in CPP/Java format"))
```
rewrite it as:
```
codemon init -n <file> <options> - - - - creates file using format given in options
```

ankushbhardwxj commented on codemon/CodemonInitJava.py:
```diff
@@ -0,0 +1,24 @@
+    # create input file
+    f = open(path + '/' + contestName + '/' + "input.txt","w+")
+    f.close()
+    print(colored.cyan('Files have been created'))
```
Move this function to `CodemonInit.py`

ankushbhardwxj commented on codemon/CodemonInitJava.py:
```diff
@@ -0,0 +1,24 @@
+      f = open(path + '/' + contestName + '/' + fileNames[files],"w+")
+      template = template_java()
+      f.write(template)
+      f.close()
```
Add a condition here to check if a template file exists at root `~/.codemon/templates`. If template file is found, use that template to write to file else display:
```
Java Template not found at ~/.codemon/templates. 
Do you wish to create one ? (Y/N)
Y     // user types Y and hits enter
Creating template file at ~/.codemon/templates ...
Successfully created template file !
```

alphaX86 committed:
Revert version to 0.0.1

alphaX86 commented:
Complete. The version is reverted @ankingcodes

ankushbhardwxj commented on codemon/codemon.py:
```diff
@@ -18,26 +20,47 @@ def main():
-        contestName = sys.argv[countArg]
-        practiceFiles = get_practice_files()
-        init(contestName, practiceFiles)
+        if sys.argv[countArg] == '-cpp':
+          contestName = sys.argv[countArg]
+          practiceFiles = get_practice_files()
+          init(contestName, practiceFiles)
+        elif sys.argv[countArg] == '-java':
+          contestName = sys.argv[countArg]
+          practiceFiles = get_practice_files_java()
```
Argument parsing is not proper ! If we type `codemon init -n newFile -java`, I get a directory created as "-java/". 
Also, `codemon init newDirectory -java` doesn't work. Same for C++ files. You need to refactor this function.

ankushbhardwxj commented on codemon/codemon.py:
```diff
@@ -18,26 +20,47 @@ def main():
+            init(contestName, fileNames)
+
+        elif sys.argv[countArg] == '-java':
+          
+          if sys.argv[countArg] == '-n':
+            file = sys.argv[countArg+1]
+            path = '.'
+            f = open(path + '/' + file + '.java',"w+")
+            template = template_java()
+            f.write(template)
+            f.close()
+            print(colored.yellow("Created "+file+'.java'))
+            break;
+          else:
+            contestName = sys.argv[countArg]
+            fileNames = get_filename_java()
+            initjava(contestName, fileNames)
 
```
You may notice that we are repeating the same stuff from Cpp part here. We don't want to duplicate code. Refactor this entire function such that we can have the least amount of code duplication here. This can be written as: 
```python
def main():

  if len(sys.argv) < 2:
    showHelp()

  else:
    countArg = 0;
    isCppFile = False
    isJavaFile = False
    isSingleFile = False
    toInit = False
    toListen = False
    toPractice = False
    targetFile = ""
    # get all arguments and flags
    for arg in sys.argv:
      countArg+=1;
      if arg == "init":
        toInit = True
      else if arg == "-cpp":
        isCppFile = True
      else if arg == "-java":
        isJavaFile = True
      else if arg == "-n":
        isSingleFile = True
      else if arg == "listen":
        isListen = True
      else if arg == "practice":
        toPractice = True
      else:
        targetFile = arg

    # based on booleans perform action
    if toListen == True:
      listen()
    else if toInit == True:
      if isJavaFile == True and isSingleFile == False:
        extension = ".java" # we can also pass it directly in function
        path = "."
        # create that directory with that given name (targetFile)
        createTargetDirectory(path, template, targetFile, extension)
      elif isJavaFile == True and isSingleFile == True:
        # extension & path passed directly here
        createTargetFile('.', template, targetFile, 'java') 
      elif isCppFile == True and isSingleFile == False:
      # and so on 
        .....
```

alphaX86 commented:
Oh wow! This many changes?! LOL. OK, I'll fix them and get back to you @ankingcodes

alphaX86 committed:
Merge branch 'master' into v0.0.2

alphaX86 committed:
Temp changes

alphaX86 committed:
Refine files

alphaX86 committed:
Cleaning up!

alphaX86 committed:
Solving errors

alphaX86 committed:
Resolving conflicts from previous commit

alphaX86 commented on codemon/codemon.py:
```diff
@@ -4,42 +4,76 @@
 
-      elif arg == "--help":
-        showHelp()
-        break
+      elif isJavaFile == True and toPractice == True:
+        name = input("Enter Contest Name: ")
+        practiceFiles = get_practice_files_java()
+        initjava(name, practiceFiles)
```
It ends here, do I have to include the listen fn @ankingcodes ?? And what else is needed?

alphaX86 commented on codemon/codemon.py:
```diff
@@ -4,42 +4,76 @@
+    if toListen == True:
+      listen()
+
+    elif toInit == True:
+      if isJavaFile == True and isSingleFile == False:
+        name = input("Enter Contest Name: ")
+        files = get_filename_java()
+        initjava(name, files)
+
+      elif isJavaFile == True and isSingleFile == True:
+        # extension & path passed directly here
+        fileName = input("Enter filename: ")
+        dirName = input("Enter directory: ")
+        file = [fileName + '.java']
+        initjava(dirName, file)
```
This time I made the tool to prompt the user to enter contest name, file names etc

alphaX86 commented on codemon/CodemonInit.py:
```diff
@@ -22,3 +22,24 @@ def init(contestName,fileNames):
     f.close()
     print(colored.cyan('Files have been created'))
+
+def initjava(contestName,fileNames):
+  # create a directory with contest name
+  try:
```
The function is moved here now

alphaX86 commented on codemon/CodemonHelp.py:
```diff
@@ -1,13 +1,11 @@
+from clint.textui import colored
+
+def showHelp():
+  print("           ---CODEMON---          ")
+  print(colored.cyan("  a CLI tool for competitive coders"))
+  print("\nCOMMANDS: \n")
+  print(colored.yellow("codemon - - - - - - - - - - - - - - - shows help"))
+  print(colored.yellow("codemon init <args>  - - - - - initialises a contest directory using format given in options"))
+  print(colored.yellow("codemon init <args> -n  - - - - - - - creates file using format given in options"))
+  print(colored.yellow("codemon listen  - - - - - - - - - - - compiles and gives output"))
+  print(colored.yellow("codemon practice <args>  - - - - - starts a practice session"))
```
Refined help menu with color codes

alphaX86 commented:
I need last thing, do I have to include listen for now? Or do I need to refine it by adding JAVA too?? These are my doubts

alphaX86 committed:
Merge https://github.com/ankingcodes/codemon into v0.0.2

alphaX86 committed:
Tuning code...

ankingcodes commented:
@alphaX86 - You don't have to review your PR to point me to things, you can add a comment with all the necessary points you want me to look into while reviewing.

alphaX86 commented:
Agreed 👍

ankingcodes commented:
@alphaX86 - I'm holding on review of this PR right now, since @paramsiddharth is adding an important feature which is going to conflict this PR. Let's merge his PR first and then you can resolve the conflicts, after which I can review and merge 👍🏻 . Thanks for your patience.

alphaX86 commented:
I understand

ankingcodes commented:
@alphaX86 - That PR got merged. You can now resolve the conflicts and ping me for review.

alphaX86 committed:
Merge branch 'master' into v0.0.2

alphaX86 committed:
Refining code and functions

alphaX86 commented:
@ankingcodes you can start the review now. As far as concerned, I ran into many issues and I fixed them with care. If there's any doubt ask me here, I'll say why I did this and that etc :)

alphaX86 committed:
Fix listen fn

alphaX86 commented:
All clear...

ankingcodes commented:
@alphaX86 - Lots of errors in your PR, also your branch is not updated. Try running `codemon init -n someFileName`, that itself gives errors. Also, please don't change colours of the outputs, revert them to how it was previously. You have created new functions which were already made in previous PR, you could've simply reused them (you couldn't because your branch isn't updated). For example, see `codemon.py` already has a function called `init_single_file` which could be reused but you created `write_to_file_single` at `CodemonInit.py`. Please look into these changes.

alphaX86 commented:
The main reason I introduced another function is that the one you've mentioned occurs many errors for me. Especially file pointer based ones. To avoid that, I introduced such. Maybe I should close this PR since this caused many issues and confusion 😟

alphaX86 commented:
The main reason I introduced another function is that the one you've mentioned occurs many errors for me. Especially file pointer based ones. To avoid that, I introduced such. Maybe I should close this PR since this caused many issues and confusion 😟

alphaX86 commented:
@ankingcodes I'm closing this PR and remake to solve #14 with proper codes.

alphaX86 commented:
> @alphaX86 - Lots of errors in your PR, also your branch is not updated. Try running `codemon init -n someFileName`, that itself gives errors. Also, please don't change colours of the outputs, revert them to how it was previously. You have created new functions which were already made in previous PR, you could've simply reused them (you couldn't because your branch isn't updated). For example, see `codemon.py` already has a function called `init_single_file` which could be reused but you created `write_to_file_single` at `CodemonInit.py`. Please look into these changes.

That's because this branch is 3 days old. Anyway, I'm closing this...
